### PR TITLE
docs: add network performance numbers

### PR DESCRIPTION
**Pull request by raduiliescu**
## Reason for This PR

For the moment there is no information about Firecracker network performance, like what is the throughput that the emulation layer is able to achieve or how much latency it adds.

## Description of Changes

Add a page with info about Firecracker network performance.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

**Comment by iulianbarbu** [comment]
Is this `performance.md` file gonna be an entry point for data about performance in general related to firecracker? At the moment it contains data only about the network performance, hence I would rename it `network-performance.md`.

**Comment by raduiliescu** [comment]
> 
> 
> Is this `performance.md` file gonna be an entry point for data about performance in general related to firecracker? At the moment it contains data only about the network performance, hence I would rename it `network-performance.md`.

My intention was to make it generic to all firecracker performance items, bootup time, network, block device. But indeed currently the first phrase is misleading.

**Comment by raduweiss** [comment]
> > Is this `performance.md` file gonna be an entry point for data about performance in general related to firecracker? At the moment it contains data only about the network performance, hence I would rename it `network-performance.md`.
> 
> My intention was to make it generic to all firecracker performance items, bootup time, network, block device. But indeed the first phrase in the doc is misleading.

Some of that is also present in [SPECIFICATION.md](https://github.com/firecracker-microvm/firecracker/blob/master/SPECIFICATION.md). Does it make sense to improve that file instead of creating a new one?

**Comment by raduiliescu** [comment]
> 
> 
> > > Is this `performance.md` file gonna be an entry point for data about performance in general related to firecracker? At the moment it contains data only about the network performance, hence I would rename it `network-performance.md`.
> > 
> > 
> > My intention was to make it generic to all firecracker performance items, bootup time, network, block device. But indeed the first phrase in the doc is misleading.
> 
> Some of that is also present in [SPECIFICATION.md](https://github.com/firecracker-microvm/firecracker/blob/master/SPECIFICATION.md). Does it make sense to improve that file instead of creating a new one?

Thanks for the suggestion! I will update that file with a summary. 

For performance tests is important to mention the environment and conditions where that numbers were taken. This way customers can build their own benchmarks and replicate. Not sure SPECIFICATION.md if that is the place to mention it. 
I suggest to add in SPECIFICATION.md something like _25Gbps when using 100% emulation thread cpu and 0.06ms of added latency_ and a link to this file with more details which I will rename as network performance.
How does this sounds?

**Comment by raduweiss** [comment]
> > Some of that is also present in [SPECIFICATION.md](https://github.com/firecracker-microvm/firecracker/blob/master/SPECIFICATION.md). Does it make sense to improve that file instead of creating a new one?
> 
> Thanks for the suggestion! I will update that file with a summary.
> 
> For performance tests is important to mention the environment and conditions where that numbers were taken. This way customers can build their own benchmarks and replicate. Not sure SPECIFICATION.md if that is the place to mention it.
> I suggest to add in SPECIFICATION.md something like _25Gbps when using 100% emulation thread cpu and 0.06ms of added latency_ and a link to this file with more details which I will rename as network performance.
> How does this sounds?

Sounds good!

Regarding the statement above: can we reach 100% emulation thread CPU usage even with small VMs (1-2 cores)? Note that since we advise all users to disable hyperthreading in production, we should maintain the same frame of reference in our performance testing/reporting.

**Comment by raduiliescu** [comment]
> 
> 
> > > Some of that is also present in [SPECIFICATION.md](https://github.com/firecracker-microvm/firecracker/blob/master/SPECIFICATION.md). Does it make sense to improve that file instead of creating a new one?
> > 
> > 
> > Thanks for the suggestion! I will update that file with a summary.
> > For performance tests is important to mention the environment and conditions where that numbers were taken. This way customers can build their own benchmarks and replicate. Not sure SPECIFICATION.md if that is the place to mention it.
> > I suggest to add in SPECIFICATION.md something like _25Gbps when using 100% emulation thread cpu and 0.06ms of added latency_ and a link to this file with more details which I will rename as network performance.
> > How does this sounds?
> 
> Sounds good!
> 
> Regarding the statement above: can we reach 100% emulation thread CPU usage even with small VMs (1-2 cores)? Note that since we advise all users to disable hyperthreading in production, we should maintain the same frame of reference in our performance testing/reporting.

We can't get 100% utilization of emulation thread using 2 cores. You need at least 4 cores to get 100% utilization.

File: docs/network-performance.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Firecracker network performance numbers
+
+This document provides details about Firecracker network performance.
+The numbers presented are dependent on the hardware (CPU, networking card, etc.), OS version and settings.
+Scope of the measurements was illustrate the limits for the emulation thread.
```
Comment (acatangiu) [inline docs/network-performance.md:5]:
```suggestion
Scope of the measurements is to illustrate the limits for the emulation thread.
```

File: docs/network-performance.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+Segment size/ Direction | 1460bytes | 256bytes | 128bytes | 96bytes
+--- | --- | --- | --- |---
+Ingress| 25Gbps | 23Gbps | 20Gbps | 18Gbps
+Egress | 25Gbps | 23Gbps | 20Gbps | 18Gbps
+Bidirectional | 18Gbps | 18Gbps | 18Gbps | 18Gbps
+
+### Setup and test description
+
+Throughput measurements were done using [iperf3](https://iperf.fr/). Scope was to make out the maximum of emulation thread and keep it in a 100% utilization.
```
Comment (acatangiu) [inline docs/network-performance.md:17]:
```suggestion
Throughput measurements were done using [iperf3](https://iperf.fr/). The target is to fully saturate the emulation thread and keep it at 100% utilization.
```

File: docs/network-performance.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+No adjustments were done to socket buffer, or any other network related kernel parameters.
+
+To identify the limit of emulation thread, TCP throughput was measured between host and guest. An EC2 [M5d.metal](https://aws.amazon.com/ec2/instance-types/m5/), instance running [Amazon Linux](https://aws.amazon.com/amazon-linux-ami/), was used as a host.
```
Comment (acatangiu) [inline docs/network-performance.md:20]:
```suggestion
To identify the limit of emulation thread, TCP throughput was measured between host and guest. An EC2 [M5d.metal](https://aws.amazon.com/ec2/instance-types/m5/) instance, running [Amazon Linux](https://aws.amazon.com/amazon-linux-ami/), was used as a host.
```

File: docs/network-performance.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+For ingress or egress throughput measurements, a Firecracker microVM running Kernel 4.14 with 4GB of Ram, 8 vCPUs, one network interface was started.
```
Comment (acatangiu) [inline docs/network-performance.md:22]:
```suggestion
For ingress or egress throughput measurements, a Firecracker microVM running Kernel 4.14 with 4GB of Ram, 8 vCPUs and one network interface was used.
```

File: docs/network-performance.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+Numbers were got using 6 iperf3 clients running on host and 6 iperf3 serves running on guest and vice versa.
```
Comment (acatangiu) [inline docs/network-performance.md:23]:
```suggestion
The measurements were taken using 6 iperf3 clients running on host and 6 iperf3 serves running on guest and vice versa.
```

File: docs/network-performance.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+For bidirectional throughput measurements a Firecracker microVM running Kernel 4.14 with 4GB of Ram, 12 vCPUs, one network interface running Amazon Linux was started.
+Numbers were got using 4 iperf3 clients and 4 iperf3 servers running on both host and guest.
```
Comment (acatangiu) [inline docs/network-performance.md:26]:
```suggestion
The measurements were taken using 4 iperf3 clients and 4 iperf3 servers running on both host and guest.
```

File: docs/network-performance.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+Throughput measurements were done using [iperf3](https://iperf.fr/). Scope was to make out the maximum of emulation thread and keep it in a 100% utilization.
+No adjustments were done to socket buffer, or any other network related kernel parameters.
+
+To identify the limit of emulation thread, TCP throughput was measured between host and guest. An EC2 [M5d.metal](https://aws.amazon.com/ec2/instance-types/m5/), instance running [Amazon Linux](https://aws.amazon.com/amazon-linux-ami/), was used as a host.
+
+For ingress or egress throughput measurements, a Firecracker microVM running Kernel 4.14 with 4GB of Ram, 8 vCPUs, one network interface was started.
+Numbers were got using 6 iperf3 clients running on host and 6 iperf3 serves running on guest and vice versa.
+
+For bidirectional throughput measurements a Firecracker microVM running Kernel 4.14 with 4GB of Ram, 12 vCPUs, one network interface running Amazon Linux was started.
```
Comment (acatangiu) [inline docs/network-performance.md:25]:
```suggestion
For bidirectional throughput measurements, a Firecracker microVM running Amazon Linux, Kernel 4.14 with 4GB of Ram, 12 vCPUs and one network interface was used.
```

File: docs/network-performance.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+### Setup and test description
+
+Latency measurements were done using ping round trip times.
+2 x EC2 M5d.metal instances running Amazon Linux within the same [VPC](https://aws.amazon.com/vpc/) were used, with a security group configured so that it would allow traffic from instances using private IPs. A 10Mbps background traffic was running between instances.
+
+Round trip time between instances was measured.
+
+```rtt min/avg/max/mdev = 0.101/0.198/0.237/0.044 ms```
+
+On one of the instances, a Firecracker microVM running Kernel 4.14, with 1 GB of RAM, 2 vCPUs, one network interface running was started.
```
Comment (acatangiu) [inline docs/network-performance.md:41]: nit:
```suggestion
On one of the instances, a Firecracker microVM running Kernel 4.14, with 1 GB of RAM, 2 vCPUs, one network interface running was used.
```

Comment (raduiliescu) [reply docs/network-performance.md]: Updated. Thanks for all the suggestions!

File: docs/network-performance.md (hunk 7 above)
Comment (raduiliescu) [reply docs/network-performance.md]: Updated

File: docs/network-performance.md (hunk 6 above)
Comment (raduiliescu) [reply docs/network-performance.md]: Updated

File: docs/network-performance.md (hunk 5 above)
Comment (raduiliescu) [reply docs/network-performance.md]: Updated

File: docs/network-performance.md (hunk 4 above)
Comment (raduiliescu) [reply docs/network-performance.md]: Updated

File: docs/network-performance.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+Ingress| 25Gbps | 23Gbps | 20Gbps | 18Gbps
+Egress | 25Gbps | 23Gbps | 20Gbps | 18Gbps
+Bidirectional | 18Gbps | 18Gbps | 18Gbps | 18Gbps
+
+### Setup and test description
+
+Throughput measurements were done using [iperf3](https://iperf.fr/). The target is to fully saturate the emulation thread and keep it at 100% utilization.
+No adjustments were done to socket buffer, or any other network related kernel parameters.
+
+To identify the limit of emulation thread, TCP throughput was measured between host and guest. An EC2 [M5d.metal](https://aws.amazon.com/ec2/instance-types/m5/) instance, running [Amazon Linux](https://aws.amazon.com/amazon-linux-ami/), was used as a host.
```
[the remark this answers is not in the document]
Comment (raduiliescu) [reply docs/network-performance.md:20]: Good catch. I'll update.

[the remark this answers is not in the document]
Comment (raduiliescu) [reply docs/network-performance.md]: Updated.

**Pull request merged**